Add new "Chain" arrow type

ZizzyZizzy opened the pull request:
When the "Chain" arrow hits a living entity, reduced damage vanilla arrows shoot out from the hit entity directly towards any other nearby living entities.

I was play-testing on 1.15.2 PaperSpigot. No errors were encountered with any of the arrow types.

ZizzyZizzy committed:
Added new "Chain" arrow type.

2008Choco reviewed (changes requested):
I'm not opposed to this change at all. In fact, this is a request I've seen come up a number of times. Aside from the two notes I've mentioned, I want to also request that you try to abide by whitespace standards as in the rest of the project. i.e. `if (shooter == hitEntity) {` in place of `if(shooter == hitEntity){`, as well as new lines to break up blocks of code where it makes the most amount of sense (generally after a closing statement).

2008Choco commented on src/main/java/wtf/choco/arrows/AlchemicalArrows.java:
```diff
@@ -122,6 +122,7 @@ public void onEnable() {
         // Register alchemical arrows
         this.getLogger().info("Registering default alchemical arrows and their recipes");
         this.createArrow(new AlchemicalArrowAir(this), "Air", Material.FEATHER); // Custom model data 132 (+1 for every registration)
+        this.createArrow(new AlchemicalArrowChain(this), "Chain", Material.FIREWORK_ROCKET);
```
Additional note I wanted to make, might there be a better item for chain arrows? I'm not sure firework rockets make a whole lot of sense here. Certainly open to accepting firework rockets as an option, just wondering if there may be an alternative.

ZizzyZizzy replied:
Open for suggestions here. I was trying to find a balance between something that made sense (firework explodes into lots of other projectiles) but not too easy to obtain.

2008Choco replied:
And that's my concern as well. The process it takes to obtain this item seems slightly too demanding in my opinion. Perhaps a slimeball is most appropriate here.

ZizzyZizzy replied:
That works too. I'll update and change colors to match.

ZizzyZizzy commented on src/main/java/wtf/choco/arrows/AlchemicalArrows.java:
```diff
@@ -65,7 +65,7 @@
 
     private static final int RESOURCE_ID = 11693;
-    private static final String SPIGET_LINK = "https://api.spiget.org/v2/resources/" + RESOURCE_ID + "/versions/latest";
+    private static final String SPIGOT_LINK = "https://api.spigot.org/v2/resources/" + RESOURCE_ID + "/versions/latest";
```
My apologies. That was indeed out of scope for the addition of the Chain arrows.

ZizzyZizzy commented on src/main/java/wtf/choco/arrows/arrow/AlchemicalArrowChain.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public void onHitPlayer(AlchemicalArrowEntity arrow, Player player) {
+        Arrow bukkitArrow = arrow.getArrow();
+        this.attemptChain(bukkitArrow, (LivingEntity) bukkitArrow.getShooter(), (LivingEntity) player);
```
Sure enough...I didn't notice it was grayed out in the IDE. Rookie mistake.

2008Choco commented on src/main/java/wtf/choco/arrows/arrow/AlchemicalArrowChain.java:
```diff
@@ -0,0 +1,78 @@
+            Vector dirToTarget = newTarget.getLocation().toVector().add(new Vector(0,newTarget.getHeight() / 2,0)).subtract(newArrowSourceLoc.toVector());
+            Arrow chainArrow = world.spawnArrow(newArrowSourceLoc,dirToTarget, 1.2F,12);
+            chainArrow.setShooter(shooter);
+            chainArrow.setDamage(chainArrow.getDamage() * 0.80);
```
Perhaps an arrow property would be a good idea here (and consequently, a configurable value). `CHAINED_DAMAGE_FACTOR`. A double value ranging from 0.0 - 1.0

ZizzyZizzy replied:
Agreed.

2008Choco commented on src/main/java/wtf/choco/arrows/arrow/AlchemicalArrowChain.java:
```diff
@@ -0,0 +1,78 @@
+    /*
+    Chain the arrow to other nearby targets.
+     */
+    private void attemptChain(Arrow source, LivingEntity shooter, LivingEntity hitEntity) {
```
Can `shooter` be made a `ProjectileSource`? I see no compelling reason for it to be a `LivingEntity`. As far as I'm concerned, this will prevent any ClassCastExceptions assuming `BlockProjectileSource` does finally get completed.

ZizzyZizzy replied:
Good catch. I wasn't even aware of BlockProjectileSource.

ZizzyZizzy committed:
Changes as requested in the pull request review.
Server owner controlled Chain damage & search radius.
Formatting changes to match existing style.
Slimeball instead of Firework for Chain arrow recipe.

ZizzyZizzy commented:
Changes pushed in fc70a66

2008Choco commented:
This all looks well and good. A few tweaks I'll make after the fact (mainly regarding whitespace), but I think this is ready to go. Thank you 😄 I appreciate the contribution. Your PR will be attributed in the next changelog.

ZizzyZizzy commented:
Awesome, thanks! It was fun. First one for a new developer here.
